src/metadata/metadata_io.c

```diff
@@ -345,140 +379,54 @@ static int metadata_io_i_asynch(ocf_cache_t cache, ocf_queue_t queue, int dir,
 	if (!a_req)
 		return -OCF_ERR_NO_MEM;
 
-	env_atomic_set(&a_req->req_remaining, io_count);
-	env_atomic_set(&a_req->req_active, io_count);
+	env_atomic_set(&a_req->req_remaining, 1);
+	env_atomic_set(&a_req->req_active, 1);
+	env_atomic_set(&a_req->req_current, -1);
 	a_req->on_complete = compl_hndl;
 	a_req->context = context;
 	a_req->page = page;
+	a_req->count = count;
 
-	/* Allocate particular requests and initialize them  */
-	OCF_REALLOC_CP(&a_req->reqs, sizeof(a_req->reqs[0]),
-			io_count, &a_req->reqs_limit);
-	if (!a_req->reqs) {
-		env_free(a_req);
-		ocf_cache_log(cache, log_warn,
-				"No memory during metadata IO\n");
-		return -OCF_ERR_NO_MEM;
-	}
 	/* IO Requests initialization */
-	for (i = 0; i < io_count; i++) {
-		env_atomic_set(&(a_req->reqs[i].req_remaining), 1);
-		env_atomic_set(&(a_req->reqs[i].finished), 0);
-		a_req->reqs[i].asynch = a_req;
+	for (i = 0; i < req_count; i++) {
+		m_req = &a_req->reqs[i];
+
+		m_req->asynch = a_req;
+		m_req->cache = cache;
+		m_req->context = context;
+		m_req->on_meta_fill = io_hndl;
+		m_req->on_meta_drain = io_hndl;
+		m_req->req.io_if = &metadata_io_restart_if;
+		m_req->req.io_queue = queue;
+		m_req->req.cache = cache;
+		m_req->req.priv = m_req;
+		m_req->req.info.internal = true;
+		m_req->req.rw = dir;
+		m_req->req.map = LIST_POISON1;
+
+		m_req->data = ctx_data_alloc(cache->owner, max_count);
```

Comment:
Please change this so that we allocate only 'count' pages here if 'count < max_count' (the case where only one slave request is allocated). Adding extra variable would do:

unsigned data_count = min(count, max_count);

Reply:
Done.